src/main/java/com/mongodb/jdbc/MongoDriver.java

```diff
@@ -99,18 +99,50 @@
 
     @Override
     public Connection connect(String url, Properties info) throws SQLException {
+        Pair<MongoConnection, Integer> p = getUnvalidatedConnectionAndTimeout(url, info);
+        Connection conn = p.left();
+        conn.isValid(p.right());
+        return conn;
+    }
+
+    protected MongoConnection createUnvalidatedConnection(String url, Properties info)
+        throws SQLException {
+        return getUnvalidatedConnectionAndTimeout(url, info).left();
+    }
+
+    private Pair<MongoConnection, Integer> getUnvalidatedConnectionAndTimeout(String url, Properties info)
+        throws SQLException {
         if (!acceptsURL(url)) {
             return null;
         }
         if (info == null) {
             info = new Properties();
         }
+
         // reuse the code getPropertyInfo to make sure the URI is properly set wrt the passed
         // Properties info value.
         Pair<ConnectionString, DriverPropertyInfo[]> p = getConnectionString(url, info);
+        // ensure that the ConnectionString and Properties are consistent.
+        reconcileProperties(p.right(), info);
+
+        ConnectionString cs = p.left();
+        Connection ret = createConnection(cs, info);
+        // use a timeout of 30s, if no timeout is specified in the URL.
+        int timeout = 30;
```

Comment:
This is not really what I had in mind when saying that we should keep the same default value if nothing is specified in the connection string.

I see that you kind of had to do this because cs.getConnectTimeout returns Null.

I think I found a way to have MongoConnection provide the settings used by the underlying Java driver, including the default values:
In MongoConnection (new code inside *** ... ***):
```
public abstract class MongoConnection implements Connection {
    **private MongoClientSettings mongoClientSettings;**
    protected MongoClient mongoClient;
    ...

    public MongoConnection(ConnectionString cs, String database, Level logLevel, File logDir) {
        Preconditions.checkNotNull(cs);
         ...
        **mongoClientSettings = MongoClientSettings.builder().applyConnectionString(cs).build();**
        mongoClient =
                MongoClients.create(
                        **mongoClientSettings**,
                        MongoDriverInformation.builder()
                                .driverName(MongoDriver.NAME)
                                .driverVersion(version)
                                .build());
         ...
    }
         
    **protected int getDefaultConnectionValidationTimeout()
    {
        return mongoClientSettings.getSocketSettings().getConnectTimeout(TimeUnit.MILLISECONDS);
    }**                       
```

Then, **getDefaultConnectionValidationTimeout** can be used directly from MongoConnection and getUnvalidatedConnectionAndTimeout does not need to return the timeout.

I would even go as far as creating a `protected IsValid()` method in MongoConnection which would use the default connectTimeout from `mongoClientSettings` making this transparent to MongoDriver.

Reply:
Yep, it's a really bad api that it throws NPE if there is no option related to the given method.